### DISCUSSION TITLE
Text on image disappear after exported to docx

### DISCUSSION THREAD

**Discussion by QiangF** [kjambunathan/org-mode-ox-odt#170] [Q&A] [answered]
The test snippet is:

    #+options:  toc:nil author:nil tex:t latex:t :H 10
    #+odt_styles_file: ./text_cn.xml

    #+caption: org unicorn
    #+attr_odt: :category "figure" :list-table t
    #+label: org
    -
      -
        #+caption[(a)]:
        [[./org-mode-unicorn.png]]
      -
        #+caption[(b)]:
        [[./org-mode-unicorn.png]]

When exported to odt format, it looks like this:
![img_2022_03_15__14_46_11](https://user-images.githubusercontent.com/8192731/158322251-29995e02-9409-45f4-a62a-2096408ec0b6.png)

But after save as docx, the subfigure lable disappears and it looks like this:
![img_2022_03_15__14_46_37](https://user-images.githubusercontent.com/8192731/158322272-be98c303-ec34-405c-a465-14b56f81a6fb.png)

The current way of putting text label on the image is incompatible with docx format.
I am looking for another ways of put text frame on top of the image.
[text_on_image.odt](https://github.com/kjambunathan/org-mode-ox-odt/files/8250812/text_on_image.odt)

**Comment by kjambunathan** [comment]
> `#+odt_styles_file: ./text_cn.xml`

I don't have access to this file.  Please create a minimal working example which I can just copy and export.

**Comment by kjambunathan** [comment]
Make sure you trim the styles to the one you actually need.

**Comment by kjambunathan** [comment]
What is your compatibility setting ...  Experiment with it.  Where are you opening `docx` ... right from LibreOffice or with MS Office.  I don't have access to any Windows machine or Microsoft Word.  So, I cannot help you there.

![Screenshot from 2022-03-15 15-48-44](https://user-images.githubusercontent.com/19869864/158356785-1d87c6aa-65c3-4997-b223-6e14bc44acf3.png)


https://help.libreoffice.org/latest/en-US/text/shared/00/00000021.html?DbPAR=SHARED#bm_id3154408

The OpenDocument format evolves over time.


| ODF version                             | Date of standard approval by OASIS                                                        | First supporting version of the software           |
| :-------------------------------------- | :---------------------------------------------------------------------------------------- | :------------------------------------------------- |
| ODF 1.0                                 | 2005-05-01                                                                                | OpenOffice.org 1.1.5 or StarOffice 7               |
| ODF 1.1                                 | 2007-02-02                                                                                | OpenOffice.org 2.2 or StarOffice 8 Update 4        |
| ODF 1.2                                 | 2011-09-30                                                                                | OpenOffice.org 3, StarOffice 9, Oracle Open Office |
| ODF 1.2 Extended (compatibility mode)   | ---                                                                                       | LibreOffice 3.5                                    |
| ODF 1.2 Extended                        | ---                                                                                       | OpenOffice.org 3.2 or StarOffice 9.2               |
| ODF 1.3                                 | TBD                                                                                       | LibreOffice 7.0                                    |
| ODF 1.3 Extended                        | ---                                                                                       | LibreOffice 7.0                                    |


In current versions, you can select to save your documents using ODF
1.2 or ODF 1.0/1.1 (for backward compatibility). Choose `Tools -
Options` `Load/Save - General`

**Comment by kjambunathan** [comment]
The  "red arrow" usually means some some text is hidden.  So, you may want to experiment with sizes or positions.

**Reply by QiangF** [reply]
You can see the incompatible frames by exporting the attached snippet.

In the image below, I resized the inner frame a little,  see the inner frame is converted to as-char image, and it is placed in a paragraph with the text.
![img_2022_03_15__21_01_39](https://user-images.githubusercontent.com/8192731/158383591-9343e543-42ee-426c-88a7-31f22d35d64f.png)
[frames.odt](https://github.com/kjambunathan/org-mode-ox-odt/files/8253479/frames.odt)
[frames.txt](https://github.com/kjambunathan/org-mode-ox-odt/files/8253480/frames.txt)

I guess the solution would be add an extra frame that contains the inner frame, the extra frame should be anchored as paragraph and wrap as run through.
The structure is like: （outer frame （extra frame （inner frame）） text ）

Actually we don't need ms office to see the docx file  is wrong, just close it and open again in libreoffice.
Although there are two freely available ms office compatibles from China. The one that works best is yozo office, but its menu is chinese only. The one with english menus is wps office
https://www.wps.com/office/linux/

**Reply by QiangF** [reply]
This is the fixed docx file, remove the .txt extension before opening. 
[frames_with_first_image_fixed.docx.txt](https://github.com/kjambunathan/org-mode-ox-odt/files/8253812/frames_with_first_image_fixed.docx.txt)

**Comment by kjambunathan** [comment, marked as answer]
>    ` #+ATTR_ODT: :outer-frame  (style "OuterFrame" anchor "as-char" extra "draw:z-index=\"1\"")`
>     `#+ATTR_ODT: :inner-frame  (style "InnerFrame" extra "draw:z-index=\"0\"")`


When LibreOffice writes to docx format it changes the paragraph anchoring of the inner frame to character anchoring.  


From [[Anchor to paragraph not saved](https://ask.libreoffice.org/t/anchor-to-paragraph-not-saved/46322)](https://ask.libreoffice.org/t/anchor-to-paragraph-not-saved/46322/4)

> See [this commit](https://gerrit.libreoffice.org/plugins/gitiles/core/+/4f40bf6a79de6d60da0a5090cdfeda6242e889f0) changing the default to "as character", because of lacking "to paragraph" in Word.

The commit itself links to the following bugs ...

1. [[Bug 45778](https://bugs.documentfoundation.org/show_bug.cgi?id=45778) - [RFE] Insert image not in a new paragraph but in the current position or as character](https://bugs.documentfoundation.org/show_bug.cgi?id=45778)
2. [[Bug 87720](https://bugs.documentfoundation.org/show_bug.cgi?id=87720) - Default insert image anchor, wrapping, and spacing](https://bugs.documentfoundation.org/show_bug.cgi?id=87720)


Some notes from the above links for later reference 


> If an image is anchored "as character", then horizontal movement and text wrap is not possible. And vertical movement will changes the line height but does not move the image to a very different position on the page. I'm sure we will get a lot of complains that "positioning image with mouse" is broken, when you will make "as character" the default.

> Word knows two states, "Inline" and "Floating". "Inline" acts the same as "as character" in LibreOffice. But default in Word is "Floating".

> The import of Word documents maps "Floating" to "to character".

> I suggest this behavior as default:
> If the cursor is in an empty paragraph, then anchor "to paragraph", and if the cursor is inside text in a paragraph, then anchor "to character".

> Besides that, bug 79160 should be fixed and the file explorer should provide an additional, special style "graphic as character".



Most of these discussions concern manual insertion of images by hand.

Regina's comment  about  "inline" and "floating" in MS Word suggests that it supports "as char" and "to char" anchors.

**Comment by kjambunathan** [comment]
> I really don't want to go down the path of creating a third frame ...  Sub-captions is something I added to make you happy.

If you look at [the margin notes](https://github.com/kjambunathan/org-mode-ox-odt/issues/169) case,  I am creating frames in the middle of running text. So, having a special block for emitting frames will be useful.  If you could give me an example odt or docx file which is well within the ODT exporter could do with minimal changes, I will look in to it.

**Reply by QiangF** [reply]
It's OK. I just hope MS office die sooner.

**Comment by kjambunathan** [comment]
> This is the fixed docx file, remove the .txt extension before opening.

> [frames_with_first_image_fixed.docx.txt](https://github.com/kjambunathan/org-mode-ox-odt/files/8253812/frames_with_first_image_fixed.docx.txt)

I think you have started with a complicated example.  It seems you have hand edited the file.  (The current ODT exporter does NOT emit `draw:custom-shape`, and the above file does have `draw:custom-shape`s)

Ultimately we want to produce something like this ...  

[frames_with_first_image_fixed.docx](https://github.com/kjambunathan/org-mode-ox-odt/files/8260929/frames_with_first_image_fixed.docx)

[frames_with_first_image_fixed.odt](https://github.com/kjambunathan/org-mode-ox-odt/files/8260930/frames_with_first_image_fixed.odt)


There are two custom shapes anchored to (the same?) char, and hence overlaid one on top of the other.    If I could find a way  out of  simplifying `draw:enhanced-geometry`, I think you may be able to achieve what you want.  Experiment a bit with the XML on your side and find out what works.  

Btw, I added some support for `draw:custom-shape` as part of https://github.com/kjambunathan/org-mode-ox-odt/blob/master/notes/SNIPPETS.org#org-odt-special-block-new-block-customshape.   I think using `draw:custom-shape` will improve compatibility with importing to Google Docs etc.  In that example, I have added only text inside the `draw:custom-shape`, and in the above documents we need to get an image within the `draw:custom-shape`.

How does the above files look at your end ...

**Reply by QiangF** [reply]
The docx file just needs one more fix: send behind text
![img_2022_03_16__21_02_26](https://user-images.githubusercontent.com/8192731/158596181-cff03b17-0da2-4813-b49a-acdb4688e9b4.png)

Or I can fix the odt file by send to backgroud.

The fixed docx file is attached here (remove .txt before open it)
[frames_with_first_image_fixed.docx.txt](https://github.com/kjambunathan/org-mode-ox-odt/files/8262318/frames_with_first_image_fixed.docx.txt)

The fixed odt file:
[frames_with_first_image_fixed.odt](https://github.com/kjambunathan/org-mode-ox-odt/files/8262362/frames_with_first_image_fixed.odt)


Now looks like:
![img_2022_03_16__21_03_14](https://user-images.githubusercontent.com/8192731/158596246-3d85e069-f0c8-4cb2-93ce-2730af27049e.png)

**Reply by QiangF** [reply]
I put the frame in table, and the first cell displayed correctly when exported to docx, but not the second cell.
![img_2022_03_16__21_35_20](https://user-images.githubusercontent.com/8192731/158601988-96f4649e-5d5e-4742-9965-80f698f44c6c.png)

[frames_in_table.odt](https://github.com/kjambunathan/org-mode-ox-odt/files/8262549/frames_in_table.odt)

If we can put custom shape like text and other drawing object like arrow, rectangle etc. on arbitrary place of an image frame, that will be super cool.

**Comment by kjambunathan** [comment]
> [Bug 94302: Filesave DOCX: Image Anchor changes from "To Paragraph" to "As Character" and image has wrong position on reopen ](https://bugs.documentfoundation.org/show_bug.cgi?id=94302)  is almost a 7-year old bug that got an update in the last month or so.  From the bug report 

> Confirmed the image on page 31 is not centered in its cell in a docx I saved.
> Image anchoring:
> ODT: Anchor to paragraph
> DOCX: Anchor as character

> "docx only has to-char (anchored) and as-char (inline), so if the request is to preserve to-para in docx, that's a wontfix."

**Comment by kjambunathan** [comment]
I have created a new issue [Provide a means "overlay" the caption over an image · Issue #172 · kjambunathan/org-mode-ox-odt](https://github.com/kjambunathan/org-mode-ox-odt/issues/172) to address some  aspects of what understood from this discussion.
